design-docs/configuration-dsl.md

```diff
@@ -0,0 +1,376 @@
+of the compiler to complete the compilation. MSVC is moving in this direction as well, and GCC also supports this workflow.
+
+The proposed solution here would be to allow users to access the tool chain resolver more directly as well as enabling them to declare custom tool
+chains. Tool chains can be located in the system PATH, at predefined locations, identified from the registry (Windows) or identified through COM
+object (Windows), fetched remotely, installed through unattended installers or package manager or manually configured.
+
+## Open Issues
+ - The source dialect & ABI is its own beast to support. It's suggested modeling it separately and use compiler version to match variant
+ - Should it be named `tool chain` (Gradle) or `toolchain` (Wikipedia)?
+ - Should the tool chain and the libraries contained within (prebuilt libraries) be two separate concepts? Meaning the toolchain resolver can be used by prebuilt dependencies but would be tied together.
+
+## DSL
+
+```
+toolChainResolvers {
+    visualStudio {
+        // additional configuration for visual studio resolver
+    }
+    myCustomResolver(MyCustomResolver) {
+        // configuration
+    }
+}
+
+toolChains {
+    myCustomGcc(Gcc) {
+        path '...'
+    }
+    myCustomVisualCpp(VisualCpp) {
+        installDir '...'
+    }
+    myCustomToolChain(ToolChain) {
+        tools.add(GccCppCompiler) {
+            it.dimensions architectures.x64, operatingSystems.windows
+            it.executable = file('/some/other/path')
+            it.withArguments { args ->
+                args << 'some-arg'
+            }
+        }
+
+        tools.add(GccLinker) {
+            it.dimensions architecture.x64, operatingSystem.windows
+            // ...
+        }
+
+        installedBy tasks.myCustomToolChainInstallTask
+
+        // TODO - Define how to add libraries
+        libraries.add(...) {
+            // added libraries can have a...
+            //   ...group, e.g. 'com.some.toolchain'
+            //   ...name, e.g. 'libc'
+            //   ...version (optional), e.g. '2.1'
+            // Adding a dependency without a version would depend on the latest or only version declared
+        }
+    }
+}
+
+repositories {
+    toolChains.myCustomToolChain
+}
+
+dependencies {
+    implementation toolChains.myCustomToolChain.libraries.libc
+    implementation visualStudio('15.0').libc
+}
+```
+
+# Build Types
+
+Build types are a conventional dimension for native binaries used to differentiate debug and release version. Additional conventional names
+are debug optimized and final. The software model, as well as Android, have this dimension as a first class citizen.
+Android typed the items inside their buildTypes container instead of relying on `ext` properties. With the new plugins, the build types will
+be typed as well.
+
+## Open Issues
+
+## DSL
```

Comment:
What is the actual impact on my project of adding a build type? Does it produce tasks with different names that I can use to build the different types? Does it change which arguments are passed to tools like the compiler, linker, etc? Does the user have control over which arguments for which tool are tied to each build type?  What would be a scenario where I would need to provide "MyDebugType" instead of using the default type supplied by Gradle?